curp/src/client_new/unary.rs

```diff
@@ -0,0 +1,590 @@
+            wait_synced_timeout: propose_timeout * 2,
+        }
+    }
+
+    /// Create a unary config
+    fn new_full(propose_timeout: Duration, wait_synced_timeout: Duration) -> Self {
+        Self {
+            propose_timeout,
+            wait_synced_timeout,
+        }
+    }
+}
+
+/// The unary client
+#[derive(Debug)]
+pub(super) struct Unary<C: Command> {
+    /// Client state
+    state: RwLock<State>,
+    /// Local server id
+    local_server_id: Option<ServerId>,
+    /// Unary config
+    config: UnaryConfig,
+    /// marker
+    phantom: PhantomData<C>,
+}
+
+impl<C: Command> Unary<C> {
+    /// Update leader
+    fn check_and_update_leader(state: &mut State, leader_id: Option<ServerId>, term: u64) -> bool {
+        match state.term.cmp(&term) {
+            Ordering::Less => {
+                // reset term only when the resp has leader id to prevent:
+                // If a server loses contact with its leader, it will update its term for election. Since other servers are all right, the election will not succeed.
+                // But if the client learns about the new term and updates its term to it, it will never get the true leader.
+                if let Some(new_leader_id) = leader_id {
+                    debug!("client term updates to {}", term);
+                    debug!("client leader id updates to {new_leader_id}");
+                    state.term = term;
+                    state.leader = Some(new_leader_id);
+                }
+            }
+            Ordering::Equal => {
+                if let Some(new_leader_id) = leader_id {
+                    if state.leader.is_none() {
+                        debug!("client leader id updates to {new_leader_id}");
+                        state.leader = Some(new_leader_id);
+                    }
+                    assert_eq!(
+                        state.leader,
+                        Some(new_leader_id),
+                        "there should never be two leader in one term"
+                    );
+                }
+            }
+            Ordering::Greater => {
+                debug!("ignore old term({}) from server", term);
+                return false;
+            }
+        }
+        true
+    }
+
+    /// Update client state based on [`FetchClusterResponse`]
+    async fn check_and_update(
+        &self,
+        res: &FetchClusterResponse,
+    ) -> Result<(), tonic::transport::Error> {
+        let mut state = self.state.write().await;
+        if !Self::check_and_update_leader(&mut state, res.leader_id, res.term) {
+            return Ok(());
+        }
+        if state.cluster_version >= res.cluster_version {
+            debug!(
+                "ignore old cluster version({}) from server",
+                res.cluster_version
+            );
+            return Ok(());
+        }
+
+        debug!("client cluster version updated to {}", res.cluster_version);
+        state.cluster_version = res.cluster_version;
```

Comment:
Please fix the cluster version issue addressed in today's meeting, that the cluster version could fall back to an older value but is still valid.

Reply:
I am modifying the implementation of the cluster version. There are issues with using a non-decreasing cluster version, so I am planning to refactor it into a hash value.

from

```txt
server: 10(old) -> 8 -> 10(new)
client: ... -> 10(old)

server will not reject the cluster version 10(old) from client.
```

to

```txt
server: #sha1 -> #sha2 -> #sha3
client: ... -> #sha1

server will only accept the cluster version #sha3.
```